exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp

```diff
@@ -0,0 +1,219 @@
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of  nor the names of its contributors may be used to
+//    endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <exotica_ilqr_solver/ilqr_solver.h>
+
+REGISTER_MOTIONSOLVER_TYPE("ILQRSolver", exotica::ILQRSolver)
+
+namespace exotica
+{
+ILQRSolver::ILQRSolver() = default;
+ILQRSolver::~ILQRSolver() = default;
+
+void ILQRSolver::SpecifyProblem(PlanningProblemPtr pointer)
+{
+    if (pointer->type() != "exotica::DynamicTimeIndexedShootingProblem")
+    {
+        ThrowNamed("This ILQRSolver can't solve problem of type '" << pointer->type() << "'!");
+    }
+    MotionSolver::SpecifyProblem(pointer);
+    prob_ = std::static_pointer_cast<DynamicTimeIndexedShootingProblem>(pointer);
+
+    // initialize Gain matrices
+    Eigen::MatrixXd x = prob_->get_X(0), u = prob_->get_U(0);
+
+    T = prob_->get_T();
+    k_gains_.assign(T, Eigen::MatrixXd(u.rows(), x.rows()));
+    ku_gains_.assign(T, Eigen::MatrixXd(u.rows(), u.rows()));
+    kv_gains_.assign(T, Eigen::MatrixXd(u.rows(), x.rows()));
+    vk_gains_.assign(T, Eigen::MatrixXd(x.rows(), x.cols()));
+
+    HIGHLIGHT_NAMED("ILQRSolver", "initialized");
+}
+
+void ILQRSolver::BackwardPass()
+{
+    const double k_min_clamp_ = -1e10;
+    const double k_max_clamp_ = 1e10;
+
+    Eigen::MatrixXd Qf = prob_->get_Qf(), R = prob_->get_R();
+    Eigen::MatrixXd X_star = prob_->get_X_star();
+    DynamicsSolverPtr dynamics_solver = prob_->GetScene()->GetDynamicsSolver();
+
+    // eq. 18
+    vk_gains_[T - 1] = Qf * dynamics_solver->StateDelta(prob_->get_X(T - 1), X_star.col(T - 1));
+    vk_gains_[T - 1] = vk_gains_[T - 1].unaryExpr([k_min_clamp_, k_max_clamp_](double x) -> double {
+        return std::min(std::max(x, k_min_clamp_), k_max_clamp_);
+    });
+
+    Eigen::MatrixXd Sk = Qf;
+    Sk = Sk.unaryExpr([k_min_clamp_, k_max_clamp_](double x) -> double {
```

Comment:
Maybe try Sk.noalias() on the left hand side of the expression

Reply:
Why? The way I understand it it's used for big expressions. What would it do here?